status/controller.go

```diff
@@ -2,53 +2,158 @@ package status
 
 import (
 	"context"
-	"fmt"
 
+	"github.com/awslabs/operatorpkg/event"
 	"github.com/awslabs/operatorpkg/object"
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	controllerruntime "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
+const (
+	MetricLabelGroup           = "group"
+	MetricLabelKind            = "kind"
+	MetricLabelNamespace       = "namespace"
+	MetricLabelName            = "name"
+	MetricLabelConditionType   = "type"
+	MetricLabelConditionStatus = "status"
+	MetricNamespace            = "operator"
+	MetricSubsystem            = "status_condition"
+)
+
 type Controller struct {
-	kubeClient client.Client
-	forObject  ConditionedObject
+	kubeClient         client.Client
+	forObject          Object
+	observedConditions map[reconcile.Request]ConditionSet
 }
 
-func NewController(client client.Client, forObject ConditionedObject) *Controller {
+func NewController(client client.Client, forObject Object) *Controller {
 	return &Controller{
-		kubeClient: client,
-		forObject:  forObject,
+		kubeClient:         client,
+		forObject:          forObject,
+		observedConditions: map[reconcile.Request]ConditionSet{},
 	}
 }
 
 func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
 	return controllerruntime.NewControllerManagedBy(m).
 		For(c.forObject).
-		Named(fmt.Sprintf("metrics.%s", object.GVK(c.forObject).Kind)).
+		Named("status").
 		Complete(c)
 }
 
 func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
-	o := c.forObject.DeepCopyObject().(ConditionedObject)
-	labels := prometheus.Labels{
-		metricLabelKind:      object.GVK(o).Kind,
-		metricLabelNamespace: req.Namespace,
-		metricLabelName:      req.Name,
+	o := c.forObject.DeepCopyObject().(Object)
+
+	objectLabels := prometheus.Labels{
+		MetricLabelGroup:     object.GVK(o).Group,
+		MetricLabelKind:      object.GVK(o).Kind,
+		MetricLabelNamespace: o.GetNamespace(),
+		MetricLabelName:      o.GetName(),
 	}
-	ConditionCount.DeletePartialMatch(labels)
 
+	ConditionCount.DeletePartialMatch(objectLabels)
```

Comment:
We're going to take a gap in our metrics if we do it like this rather than updating if the objectLabels already exist and deleting the metric if the object labels don't exist.

It scares me a bit to do things this way since we don't persist data if our calls to the apiserver start to fail

Reply:
There's an idea of how this might be handled that's present in a metric family implementation in `kubernetes-sigs/karpenter`: https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/metrics/store.go#L25

Reply:
Nice ref, will play with this.